### ISSUE TITLE
App errors out when given an in-progress run (a run with no "result" value)

### ISSUE THREAD

**Issue by DeeDeeG** [autumnblazey/atomcommunity-pipelines#1]
Hi, 

First, thanks for making this really neat dashboard!

And second, I had a small bug to report.

---

I ran into a situation where the dashboard won't display anything.

Digging into the error console, I found this: When there is an in-progress run over at Azure DevOps, and the run is not marked "complete", apparently that run won't have a "result" value. In that case, in this web app there is a type check that fails, where it expects a string, but it gets `undefined` instead.

I grabbed a copy of the input data that caused this error:

<details><summary>input.json (click to expand)</summary>

```json
{
  "count": 12,
  "value": [
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1290"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1290"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-30T08:31:59.6828058Z",
      "finishedDate": "2022-06-30T12:11:14.4638035Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1290",
      "id": 1290,
      "name": "1.63.0-dev+1290"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1284"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1284"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "inProgress",
      "createdDate": "2022-06-30T05:37:38.8670972Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1284",
      "id": 1284,
      "name": "1.63.0-dev+1284"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1183"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1183"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-22T00:23:55.6273699Z",
      "finishedDate": "2022-06-22T01:45:20.0795482Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1183",
      "id": 1183,
      "name": "1.63.0-dev+1183"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1174"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1174"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-21T12:41:14.8932702Z",
      "finishedDate": "2022-06-21T13:58:21.6051657Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1174",
      "id": 1174,
      "name": "1.63.0-dev+1174"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1169"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1169"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-20T22:46:25.2030863Z",
      "finishedDate": "2022-06-20T23:44:21.0214022Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1169",
      "id": 1169,
      "name": "1.63.0-dev+1169"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1168"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1168"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-20T22:21:39.9328171Z",
      "finishedDate": "2022-06-20T23:25:46.1085758Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1168",
      "id": 1168,
      "name": "1.63.0-dev+1168"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1165"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1165"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-19T23:36:41.3250938Z",
      "finishedDate": "2022-06-20T00:27:57.9175919Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1165",
      "id": 1165,
      "name": "1.63.0-dev+1165"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1164"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1164"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-19T23:16:15.303536Z",
      "finishedDate": "2022-06-20T00:37:20.389024Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1164",
      "id": 1164,
      "name": "1.63.0-dev+1164"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1153"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1153"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-15T21:04:32.6774419Z",
      "finishedDate": "2022-06-19T14:26:04.5004422Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1153",
      "id": 1153,
      "name": "1.63.0-dev+1153"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1152"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1152"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-15T21:03:34.0235935Z",
      "finishedDate": "2022-06-15T22:45:09.2564225Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1152",
      "id": 1152,
      "name": "1.63.0-dev+1152"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1149"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1149"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "failed",
      "createdDate": "2022-06-11T04:35:20.8113932Z",
      "finishedDate": "2022-06-11T05:30:41.4578878Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1149",
      "id": 1149,
      "name": "1.63.0-dev+1149"
    },
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1070"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1070"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "completed",
      "result": "succeeded",
      "createdDate": "2021-04-08T01:09:17.4891105Z",
      "finishedDate": "2021-04-09T02:43:54.3061613Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1070",
      "id": 1070,
      "name": "1.57.0-dev+1070"
    }
  ]
}
```

</details>

Particularly this run:

<details><summary>In-progress_run.json (click to expand)</summary>

```json
    {
      "_links": {
        "self": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1284"
        },
        "web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/results?buildId=1284"
        },
        "pipeline.web": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_build/definition?definitionId=10"
        },
        "pipeline": {
          "href": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3"
        }
      },
      "pipeline": {
        "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10?revision=3",
        "id": 10,
        "revision": 3,
        "name": "Release Branch Build",
        "folder": "\\atom-community"
      },
      "state": "inProgress",
      "createdDate": "2022-06-30T05:37:38.8670972Z",
      "url": "https://dev.azure.com/atomcommunity/2179356e-4464-463a-8aea-5ddc7d857fda/_apis/pipelines/10/runs/1284",
      "id": 1284,
      "name": "1.63.0-dev+1284"
    },
```

</details>

Given this example data structure, where there can be runs with no "result" value, I wonder if it would be possible to update the code to handle that kind of data as representing a valid run.

**Edit to add the actual error:**

<details><summary>errors.json (click to expand)</summary>

```json
Object { count: 12, value: (12) […] }
entry-7df98a7c.mjs:1:125664
ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "value",
              1,
              "result"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "value",
              1,
              "result"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "value",
      1,
      "result"
    ],
    "message": "Invalid input"
  }
]
    NuxtJS 29
entry-7df98a7c.mjs:1:11726
    NuxtJS 22
TypeError: can't access property "value", r.runs is undefined
    NuxtJS 26
entry-7df98a7c.mjs:1:11726
```

</details>

---

**In summary:** The app expects a "result" value, but in-progress runs don't provide that (see example data above). Maybe the app can be updated to handle these in-progress runs?

Thanks again for making this, hope you don't mind the bug report!

**Comment by meadowsys** [comment]
Ah this was something I did not account for, since at the time of writing this there was not any in progress runs running. Thanks for the report >~<
